ecs-deploy

```diff
@@ -1,408 +1,466 @@
+      TASK_DEFINITION=`$AWS_ECS describe-task-definition --task-def $TASK_DEFINITION_ARN`
+    fi
+}
 
-echo "Current task definition: $TASK_DEFINITION";
-
-# Get a JSON representation of the current task definition
-# + Update definition to use new image name
-# + Filter the def
-DEF=$( $AWS_ECS describe-task-definition --task-def $TASK_DEFINITION \
-        | sed -e "s|\"image\": *\"${imageWithoutTag}:.*\"|\"image\": \"${useImage}\"|g" \
-        | sed -e "s|\"image\": *\"${imageWithoutTag}\"|\"image\": \"${useImage}\"|g" \
-        | jq '.taskDefinition' )
-
-# Default JQ filter for new task definition
-NEW_DEF_JQ_FILTER="family: .family, volumes: .volumes, containerDefinitions: .containerDefinitions"
-
-# Some options in task definition should only be included in new definition if present in
-# current definition. If found in current definition, append to JQ filter.
-CONDITIONAL_OPTIONS=(networkMode taskRoleArn)
-for i in "${CONDITIONAL_OPTIONS[@]}"; do
-  re=".*${i}.*"
-  if [[ "$DEF" =~ $re ]]; then
-    NEW_DEF_JQ_FILTER="${NEW_DEF_JQ_FILTER}, ${i}: .${i}"
-  fi
-done
-
-# Build new DEF with jq filter
-NEW_DEF=$(echo $DEF | jq "{${NEW_DEF_JQ_FILTER}}")
-
-# Register the new task definition, and store its ARN
-NEW_TASKDEF=`$AWS_ECS register-task-definition --cli-input-json "$NEW_DEF" | jq -r .taskDefinition.taskDefinitionArn`
-echo "New task definition: $NEW_TASKDEF";
-
-if [ $SERVICE == false ]; then
-  echo "Task definition updated successfully"
-else
-  DEPLOYMENT_CONFIG=""
-  if [ $MAX != false ]; then
-    DEPLOYMENT_CONFIG=",maximumPercent=$MAX"
-  fi
-  if [ $MIN != false ]; then
-    DEPLOYMENT_CONFIG="$DEPLOYMENT_CONFIG,minimumHealthyPercent=$MIN"
-  fi
-  if [ ! -z "$DEPLOYMENT_CONFIG" ]; then
-    DEPLOYMENT_CONFIG="--deployment-configuration ${DEPLOYMENT_CONFIG:1}"
-  fi
-
-  DESIRED_COUNT=""
-  if [ ! -z ${DESIRED+undefined-guard} ]; then
-    DESIRED_COUNT="--desired-count $DESIRED"
-  fi
-
-  # Update the service
-  UPDATE=`$AWS_ECS update-service --cluster $CLUSTER --service $SERVICE $DESIRED_COUNT --task-definition $NEW_TASKDEF $DEPLOYMENT_CONFIG`
-
-  # Only excepts RUNNING state from services whose desired-count > 0
-  SERVICE_DESIREDCOUNT=`$AWS_ECS describe-services --cluster $CLUSTER --service $SERVICE | jq '.services[]|.desiredCount'`
-  if [ $SERVICE_DESIREDCOUNT -gt 0 ]; then
-    # See if the service is able to come up again
-    every=10
-    i=0
-    while [ $i -lt $TIMEOUT ]
-    do
-      # Scan the list of running tasks for that service, and see if one of them is the
-      # new version of the task definition
+function createNewTaskDefJson() {
+    # Get a JSON representation of the current task definition
+    # + Update definition to use new image name
+    # + Filter the def
+    DEF=$( echo "$TASK_DEFINITION" \
+            | sed -e "s|\"image\": *\"${imageWithoutTag}:.*\"|\"image\": \"${useImage}\"|g" \
```

Comment:
We can make changes to make it more robust or stable, this round though I was just restructuring it for functions and testing. But once it is merged in if you have other improvements I'm happy to incorporate them.